test/functional/wallet_taproot.py

```diff
@@ -0,0 +1,272 @@
+        "xpub": "tpubD6NzVbkrYhZ4WqYFvnJ3Vyw5TrpME8jLf3zbd1DvKbX7jbwc5wewYLKLSFRzZWV6hZj7XhsXAy7fhE5jB25DiWyNM3ztXbsXHRVCrp5BiPY",
+        "pubs": [
+            "2258b1c3160be0864a541854eec9164a572f094f7562628281a8073bb89173a7",
+            "83df59d0a5c951cdd62b7ab225a62079f48d2a333a86e66c35420d101446e92e",
+            "2a654bf234d819055312f9ca03fad5836f9163b09cdd24d29678f694842b874a",
+            "aa0334ab910047387c912a21ec0dab806a47ffa38365060dbc5d47c18c6e66e7",
+        ]
+    },
+    {
+        "xprv": "tprv8mGPkMVz5mZuJDnC2NjjAv7E9Zqa5LCgX4zawbZu5nzTtLb5kGhPwycX4H1gtW1f5ZdTKTNtQJ61hk71F2TdcQ93EFDTpUcPBr98QRji615",
+        "xpub": "tpubDHxRtmYEE9FaBgoyv2QKaKmLibMWEfPb6NbNE7cCW4nripqrNfWz8UEPEPbHCrakwLvwFfsqoaf4pjX4gWStp4nECRf1QwBKPkLqnY8pHbj",
+        "pubs": [
+            "00a9da96087a72258f83b338ef7f0ea8cbbe05da5f18f091eb397d1ecbf7c3d3",
+            "b2749b74d51a78f5fe3ebb3a7c0ff266a468cade143dfa265c57e325177edf00",
+            "6b8747a6bbe4440d7386658476da51f6e49a220508a7ec77fe7bccc3e7baa916",
+            "4674bf4d9ebbe01bf0aceaca2472f63198655ecf2df810f8d69b38421972318e",
+        ]
+    }
+]
+
+CHANGE_XPRV = "tprv8ZgxMBicQKsPcyDrWwiecVnTtFmfRwbfFqEfR4ZGWvq5aTTwLBWmAm5zrbMcYtb9gQNFfhRfqhhrBG37U3nhmXxEgeEPBJGHAPrHCrAd1WX"
+CHANGE_XPUB = "tpubD6NzVbkrYhZ4WSFeQbPF1uSaTHHbbGnZq8qShabZwCdUQwihxaLMMFhs2kidGF2qrRKiQVqw8VoyuTHj1bZqmMXMeciaU1gBjWA1sim2zUB"
+
+# Point with no known discrete log.
+H_POINT = "50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0"
+
+
+def key(hex_key):
+    """Construct an x-only pubkey from its hex representation."""
+    return bytes.fromhex(hex_key)
+
+def pk(hex_key):
+    """Construct a script expression for taproot_construct for pk(hex_key)."""
+    return (None, CScript([bytes.fromhex(hex_key), OP_CHECKSIG]))
+
+def compute_taproot_address(pubkey, scripts):
+    """Compute the address for a taproot output with given inner key and scripts."""
+    tap = taproot_construct(pubkey, scripts)
+    assert tap.scriptPubKey[0] == 0x51
+    assert tap.scriptPubKey[1] == 0x20
+    return encode_segwit_address("bcrt", 1, tap.scriptPubKey[2:])
+
+class WalletTaprootTest(BitcoinTestFramework):
+    """Test generation and spending of P2TR address outputs."""
+
+    def set_test_params(self):
+        self.num_nodes = 2
+        self.setup_clean_chain = True
+        self.extra_args = [['-keypool=100'], ['-keypool=100']]
+        self.supports_cli = False
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+        self.skip_if_no_sqlite()
+
+    def setup_network(self):
+        self.setup_nodes()
+
+    def init_wallet(self, i):
+        pass
+
+    @staticmethod
+    def rand_keys(n):
+        ret = []
+        idxes = set()
+        for _ in range(n):
+            while True:
+                i = random.randrange(len(KEYS))
+                if not i in idxes:
+                    break
+            idxes.add(i)
+            ret.append(KEYS[i])
+        return ret
```

Comment:
I think this can be replaced with `set(random.sample(KEYS, k=n))`

Reply:
I think you're right.